internal/gatewayapi/validate.go

```diff
@@ -260,6 +268,80 @@ func (t *Translator) validateBackendServiceImport(backendRef *gwapiv1a2.BackendR
 	return true
 }
 
+func (t *Translator) validateBackendRefBackend(backendRef *gwapiv1a2.BackendRef, parentRef *RouteParentContext, resources *Resources,
+	backendNamespace string, route RouteContext, kind gwapiv1.Kind,
+) bool {
+	// TODO: support additional route kinds
+	routeStatus := GetRouteStatus(route)
+
+	if !t.BackendEnabled {
+		status.SetRouteStatusCondition(routeStatus,
+			parentRef.routeParentStatusIdx,
+			route.GetGeneration(),
+			gwapiv1.RouteConditionResolvedRefs,
+			metav1.ConditionFalse,
+			gwapiv1.RouteReasonUnsupportedValue,
+			fmt.Sprintf("Resource %s/%s of type Backend cannot be used since Backend is disabled in Envoy Gateway configuration",
+				NamespaceDerefOr(backendRef.Namespace, route.GetNamespace()), string(backendRef.Name)),
+		)
+		return false
+	}
+
+	if kind != KindHTTPRoute {
+		status.SetRouteStatusCondition(routeStatus,
+			parentRef.routeParentStatusIdx,
+			route.GetGeneration(),
+			gwapiv1.RouteConditionResolvedRefs,
+			metav1.ConditionFalse,
+			gwapiv1.RouteReasonUnsupportedValue,
+			fmt.Sprintf("Resource %s/%s of type Backend is not supported for %s routes", NamespaceDerefOr(backendRef.Namespace, route.GetNamespace()),
+				string(backendRef.Name), kind),
+		)
+		return false
+	}
+
+	backend := resources.GetBackend(backendNamespace, string(backendRef.Name))
+	if backend == nil {
+		status.SetRouteStatusCondition(routeStatus,
+			parentRef.routeParentStatusIdx,
+			route.GetGeneration(),
+			gwapiv1.RouteConditionResolvedRefs,
+			metav1.ConditionFalse,
+			gwapiv1.RouteReasonBackendNotFound,
+			fmt.Sprintf("Backend %s/%s not found", NamespaceDerefOr(backendRef.Namespace, route.GetNamespace()),
+				string(backendRef.Name)),
+		)
+		return false
+	}
+
+	var unsupportedRef bool
+
+	for _, bep := range backend.Spec.Endpoints {
+		switch {
+		case bep.Unix != nil:
+			unsupportedRef = true
+		case bep.FQDN != nil && bep.FQDN.Hostname == "localhost":
```

Comment:
If the intent is to prevent localhost from being the target, in case of an FQDN the better approach would be to try to resolve the hostname and see that it doesn't resolve to the localhost.

Reply:
What would happen if somebody were to use an IP in the `FQDN` field of `BackendEndpoint`? I think it would just work, except that it would be possible to use `127.0.0.1` here.

Reply:
> If the intent is to prevent localhost from being the target, in case of an FQDN the better approach would be to try to resolve the hostname and see that it doesn't resolve to the localhost. 

Users can change the DNS record this after the resource is accepted. Then, it becomes a race with the next translation until the config is rejected. So, it's still possible that there's a time window where localhost routing is possible. 

We touched this point here: https://github.com/envoyproxy/gateway/pull/3063#discussion_r1597285050. I'm not sure that there really is a good solution for that. The compromise is to enforce validations equivalent to what K8s allows in EP slices.

Reply:
> What would happen if somebody were to use an IP in the FQDN field of BackendEndpoint? 

There's a CEL validation that will probably reject IPv6 Addresses (as ":" is nto supported). But, IPv4 addresses are actually accepted by this validation. I'll add additional programatic validations.  
https://github.com/envoyproxy/gateway/blob/7f2038fdba662c2de9a3fb4139800c24ecc875b6/api/v1alpha1/backend_types.go#L103